smartsim/entity/application.py

```diff
@@ -0,0 +1,288 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import copy
+import typing as t
+from os import path as osp
+
+from .._core.utils.helpers import expand_exe_path
+from ..log import get_logger
+from .entity import SmartSimEntity
+from .files import EntityFiles
+
+if t.TYPE_CHECKING:
+    from smartsim.types import TODO
+
+    RunSettings = TODO
+    BatchSettings = TODO
+
+
+logger = get_logger(__name__)
+
+
+# TODO: Remove this supression when we strip fileds/functionality
+#       (run-settings/batch_settings/params_as_args/etc)!
+# pylint: disable-next=too-many-public-methods
+
+
+class Application(SmartSimEntity):
+    def __init__(
+        self,
+        name: str,
+        exe: str,
+        exe_args: t.Optional[t.Union[str, t.Sequence[str]]] = None,
+        files: t.Optional[EntityFiles] = None,
+        file_parameters: t.Mapping[str, str] | None = None,
+    ) -> None:
+        """Initialize an ``Application``
+
+        :param name: name of the application
+        :param exe: executable to run
+        :param exe_args: executable arguments
+        :param files: files to be copied, symlinked, and/or configured prior to
+                      execution
+        :param file_parameters: parameters and values to be used when configuring
+                                files
+        """
+        super().__init__(name)
+        self._exe = expand_exe_path(exe)
```

Comment:
I think the group had agreed to put doc strings after class attributes to enable intellisense and documentation.  Should we add these to the ``Application`` class?

Reply:
added